actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -822,7 +822,7 @@ def url_for(options, route_name = nil, url_strategy = UNKNOWN, method_name = nil
         path = route_with_params.path(method_name)
         params = route_with_params.params
 
-        if options[:params].is_a?(Hash)
+        if options[:params].is_a?(Hash) || options[:params].is_a?(ActionController::Parameters)
```

Comment:
I suggest making this even more open, to anything that responds to merge

```suggestion
        if options[:params].respond_to? :merge!
```